crates/iceberg/src/input.rs

```diff
@@ -0,0 +1,886 @@
+};
+use dbsp::circuit::tokio::TOKIO;
+use feldera_adapterlib::{
+    catalog::{ArrowStream, InputCollectionHandle},
+    errors::metadata::ControllerError,
+    format::ParseError,
+    transport::{
+        InputConsumer, InputEndpoint, InputQueue, InputReader, InputReaderCommand,
+        IntegratedInputEndpoint, NonFtInputReaderCommand,
+    },
+    utils::datafusion::{
+        execute_query_collect, execute_singleton_query, timestamp_to_sql_expression,
+        validate_sql_expression, validate_timestamp_column,
+    },
+    PipelineState,
+};
+use feldera_types::{
+    program_schema::Relation,
+    transport::iceberg::{IcebergCatalogType, IcebergReaderConfig},
+};
+use futures_util::StreamExt;
+use iceberg::{io::FileIO, spec::TableMetadata, table::Table as IcebergTable, Catalog, TableIdent};
+use iceberg_catalog_glue::{
+    GlueCatalog, GlueCatalogConfig, AWS_ACCESS_KEY_ID, AWS_PROFILE_NAME, AWS_REGION_NAME,
+    AWS_SECRET_ACCESS_KEY, AWS_SESSION_TOKEN,
+};
+use iceberg_catalog_rest::{RestCatalog, RestCatalogConfig};
+use iceberg_datafusion::IcebergTableProvider;
+use log::{debug, info, trace};
+use std::sync::Arc;
+use tokio::{
+    select,
+    sync::{
+        mpsc,
+        watch::{channel, Receiver, Sender},
+    },
+};
+
+// /// Polling interval when following a delta table.
+// const POLL_INTERVAL: Duration = Duration::from_millis(1000);
+
+// /// Polling delay before retrying an unsuccessful read from a delta log.
+// const RETRY_INTERVAL: Duration = Duration::from_millis(2000);
+
+enum SnapshotDescr {
+    /// Open the latest snapshot (default)
+    Latest,
+    /// Open specific snapshot id.
+    SnapshotId(i64),
+    /// Open
+    Timestamp(DateTime<Utc>),
+}
+
+/// Integrated input connector that reads from an Iceberg table.
+pub struct IcebergInputEndpoint {
+    inner: Arc<IcebergInputEndpointInner>,
+}
+
+impl IcebergInputEndpoint {
+    pub fn new(
+        endpoint_name: &str,
+        config: &IcebergReaderConfig,
+        consumer: Box<dyn InputConsumer>,
+    ) -> Self {
+        Self {
+            inner: Arc::new(IcebergInputEndpointInner::new(
+                endpoint_name,
+                config.clone(),
+                consumer,
+            )),
+        }
+    }
+}
+
+impl InputEndpoint for IcebergInputEndpoint {
+    fn is_fault_tolerant(&self) -> bool {
+        false
+    }
+}
+
+impl IntegratedInputEndpoint for IcebergInputEndpoint {
+    fn open(&self, input_handle: &InputCollectionHandle) -> AnyResult<Box<dyn InputReader>> {
+        Ok(Box::new(IcebergInputReader::new(
+            &self.inner,
+            input_handle,
+        )?))
+    }
+}
+
+struct IcebergInputReader {
+    sender: Sender<PipelineState>,
+    inner: Arc<IcebergInputEndpointInner>,
+}
+
+impl IcebergInputReader {
+    fn new(
+        endpoint: &Arc<IcebergInputEndpointInner>,
+        input_handle: &InputCollectionHandle,
+    ) -> AnyResult<Self> {
+        // TODO: perform validation as part of deserialization.
```

Comment:
config deserialization?

Reply:
yes, this way we'll get the error at compilation time.